File: AccessEye/AI/ModelStore.swift
//
//  ModelStore.swift
//  AccessEye
//
//  Knows WHERE the on-device model file lives and how to install it. The model
//  is too large to ship in the App Store binary, so it's downloaded once on
//  first launch and kept in Application Support (excluded from iCloud backup so
//  it doesn't bloat the user's backup). (README §3 Risk #2.)
//
//  Both `ModelManager` (downloads it) and `GemmaService` (loads it) go through
//  here, so there is a single source of truth for the path.
//

import Foundation

enum ModelStore {

    /// Directory holding the model, created on demand.
    static var directory: URL {
        let base = FileManager.default.urls(for: .applicationSupportDirectory,
                                            in: .userDomainMask)[0]
        return base.appendingPathComponent("Model", isDirectory: true)
    }

    /// Final on-device location of the model file.
    static var fileURL: URL {
        directory.appendingPathComponent(AppConfig.modelFileName)
    }

    /// True once the model has been fully downloaded and installed.
    static var isInstalled: Bool {
        FileManager.default.fileExists(atPath: fileURL.path)
    }

    /// True when the only model available is one baked into the app bundle (dev
    /// builds). Such a model is read-only and cannot be deleted.
    static var isBundledOnly: Bool {
        !isInstalled && usableModelPath != nil
    }

    /// Size of the usable model on disk in bytes (downloaded or bundled; 0 if none).
    static var installedSizeBytes: Int64 {
        guard let path = usableModelPath else { return 0 }
        let values = try? URL(fileURLWithPath: path)
            .resourceValues(forKeys: [.fileSizeKey, .totalFileAllocatedSizeKey])
        return Int64(values?.totalFileAllocatedSize ?? values?.fileSize ?? 0)
    }

    /// A human-readable size string for the installed model.
    static var installedSizeText: String {
        ByteCountFormatter.string(fromByteCount: installedSizeBytes, countStyle: .file)
    }

    /// A path to a usable model: the downloaded copy if present, otherwise a
    /// copy bundled in the app (handy for dev builds — drag the model into the
    /// target and it's found without downloading). Returns nil if neither.
    static var usableModelPath: String? {
        if isInstalled { return fileURL.path }
        // Dev fallback: model dragged into the app bundle.
        for ext in ["litertlm", "task", "bin"] {
            let name = (AppConfig.modelFileName as NSString).deletingPathExtension
            if let path = Bundle.main.path(forResource: name, ofType: ext) {
                return path
            }
            if let path = Bundle.main.path(forResource: "model", ofType: ext) {
                return path
            }
        }
        return nil
    }

    /// Move a freshly downloaded temp file into its final location atomically.
    static func install(from tempURL: URL) throws {
        let fm = FileManager.default
        try fm.createDirectory(at: directory, withIntermediateDirectories: true)
        if fm.fileExists(atPath: fileURL.path) {
            try fm.removeItem(at: fileURL)
        }
        try fm.moveItem(at: tempURL, to: fileURL)
        excludeFromBackup(fileURL)
    }

    /// Remove the downloaded model (e.g. to free space or re-download).
    static func remove() {
        try? FileManager.default.removeItem(at: fileURL)
    }

    private static func excludeFromBackup(_ url: URL) {
        var url = url
        var values = URLResourceValues()
        values.isExcludedFromBackup = true
        try? url.setResourceValues(values)
    }
}

File: android/app/src/main/java/gr/orestislef/accesseye/ai/ModelStore.kt
//
//  ModelStore.kt
//  AccessEye
//
//  Knows WHERE the on-device model file lives and how to install it. The model
//  is too large to ship in the Play Store binary, so it's downloaded once on
//  first launch and kept in the app's no-backup files directory (excluded from
//  Auto Backup by location so it doesn't bloat the user's backup — mirrors the
//  iOS isExcludedFromBackup flag). (README §3 Risk #2.)
//
//  Both `ModelManager` (downloads it) and `GemmaService` (loads it) go through
//  here, so there is a single source of truth for the path.
//

package gr.orestislef.accesseye.ai

import android.content.Context
import android.text.format.Formatter
import java.io.File
import java.io.IOException

class ModelStore(context: Context) {

    // Formatter.formatFileSize needs a Context, so keep the application one.
    private val appContext: Context = context.applicationContext

    /** Directory holding the model, created on demand. */
    val directory: File = File(appContext.noBackupFilesDir, "model")

    /** Final on-device location of the model file. */
    val file: File = File(directory, AppConfig.modelFileName)

    /** In-progress download location (supports resume across launches). */
    val partFile: File = File(directory, AppConfig.modelFileName + ".part")

    /** True once the model has been fully downloaded and installed. */
    val isInstalled: Boolean
        get() = file.exists()

    /** Size of the installed model on disk in bytes (0 if none). */
    val installedSizeBytes: Long
        get() = if (isInstalled) file.length() else 0L

    /** A human-readable size string for the installed model. */
    val installedSizeText: String
        get() = Formatter.formatFileSize(appContext, installedSizeBytes)

    /** Move a freshly downloaded temp file into its final location atomically. */
    fun install(temp: File) {
        directory.mkdirs()
        if (file.exists() && !file.delete()) {
            throw IOException("Could not replace the existing model file.")
        }
        if (!temp.renameTo(file)) {
            throw IOException("Could not move the downloaded model into place.")
        }
    }

    /** Remove the downloaded model (e.g. to free space or re-download). */
    fun remove() {
        file.delete()
        // Also drop any stale partial download so the next attempt starts clean.
        partFile.delete()
    }
}
